Documentation/design-docs/corefx-testing-framework.md

```diff
@@ -0,0 +1,53 @@
+
+## Testing pipeline
+
+Once the necessary components of the test host have been restored or built, a set of pre-built CoreFX binaries is downloaded from the urls specified in TestList.json found in the URL defined in `tests\CoreFX\CoreFXTestListURL[_Linux|_OSX].txt`. Once downloaded the tests are unarchived into each test folders' `xunit.console.netcore` executable is run with the produced `dotnet` CLI instance in the test host. To be able to exclude failing tests from running, the testing pipeline parses the tests defined `TopN.CoreFX.x64.issues.json` and creates a `.rsp` file, the path to which is passed as a command-line argument to `xunit.console.netcore`.
+The test scripts start and run multiple instances of the xunit runner in parallel. The test results are output under the `bin\Logs` folder and laterparsed by the CI machines.
+
+## Test binaries
+
+The binaries used to run the tests are stored in Azure Blob Storage.
+The current layout of blob storage is as follows – the root of the blob contains three folders for each OS. Each contains archived test binaries (i.e. the output of running `build-test` in CoreFX) and a .json file with a list of tests and their respective URLs. The json schema follows Helix's test list definitions, so that the test runs are compatible with Helix URLs.  
+
+To update them you need access to the secure connection string. Once you have the connection string available, you can use either Azure CLI, Azure Storage Explorer or the Blob Storage API.
+
+# Limitations and Future Development
+
+## Full architecture/OS matrix support
+
+This is the big lacking feature at the moment.
+
+## Exclusion list
+
+Defining all exclusions in a single file should be fairly straightforward. The schema is meant to be easily extensible - adding fields for exclusions on specific architectures and platforms would require a small change in how we generate `.rsp` files for the xunit test runner.
+
+## Blob Storage
+
+The current layout of blob storage is inelegant and not easy to support. We can define two potential paths for improvement:
+
++ Remove the .txt files with test URLs. Add a simple request handler between the blob storage and the testing framework to resolve URLs automatically for every supported platform/OS combo. This is where the binary refresh logic can also live – i.e. every time we update the CoreFX reference in CoreCLR, maestro-bot can send a request to refresh binaries. Then we can use helper scripts to generate test exclusions from failed tests so that CI is green while bugs are triaged.
+
++ The blob storage features above would duplicate some Helix functionality. The test list in each blob storage folder copies Helix’s JSON format. This was done intentionally to be compatible with the produced test payloads. We can use a Helix API to integrate test binary refresh with something similar to dotnet-maestro-bot and copy test binaries from a Helix run on a regular interval.
```

Comment:
Another point related to Helix. We are not taking advantage of Helix parallelization capabilities right now (all tests run on a single machine in the CI). It would be nice if the CI runs were scheduled using Helix same way as they are scheduled in the CoreFX repo. This would make the wallclock time to finish the CoreFX leg much shorter. I do not think it is high priority at the moment, but it may become a lot more interesting once we add more expensive stress variants of the tests.